src/java.base/share/classes/java/lang/ref/Reference.java

```diff
@@ -511,29 +540,32 @@ protected Object clone() throws CloneNotSupportedException {
     }
 
     /**
-     * Ensures that the object referenced by the given reference remains
+     * Ensures that the given object remains
      * <a href="package-summary.html#reachability"><em>strongly reachable</em></a>,
-     * regardless of any prior actions of the program that might otherwise cause
-     * the object to become unreachable; thus, the referenced object is not
+     * regardless of any other actions of the program that might otherwise cause
+     * the object to become unreachable; thus, the object is not
      * reclaimable by garbage collection at least until after the invocation of
-     * this method.  Invocation of this method does not itself initiate garbage
-     * collection or finalization.
+     * this method. References to the given object will not be cleared (or
+     * enqueued, if applicable) by the garbage collector until after invocation
+     * of this method.
+     * Invocation of this method does not itself initiate reference processing,
+     * garbage collection, or finalization.
```

Comment:
My understanding was that it is not the object instance that is being guarded, only that the reference holding the object is considered a strong root and is only used to delimit a range of bytecodes for which the reference is considered to be strong.
In particular, the invocation of the method itself has no semantics, only that a control flow could reach that statement and the reference was considered strong as long as the reference was in a scope that included the reachability fence.

Reply:
IMO, the core of the semantics here are that the reachabilityFence() call happens before clearing or enqueueing of any References to the argument. I think it's the call itself, not the end of the argument.s scope (which may not even make sense) that matters here.

Perhaps reachabilityFence(x) should also happen before any objects that are definitely reachable from x (JLS 12.6.2) at that point are cleared or enqueued. That is not completely clear to me, since it impacts the legality of dead field elimination. And it would be nice to get rid of 12.6.2.

Reply:
The existing docs refer to "invocation" of this method. I've continued with that, and in general have kept this at a bit higher level, in order to simplify understanding.

I'm open to more detailed wording, if it would improve understandability, (or if what I have is not functionally accurate).

Reply:
> Perhaps reachabilityFence(x) should also happen before any objects that are definitely reachable from x (JLS 12.6.2) at that point are cleared or enqueued.

I don't think we need to extend this to objects reachable from x.

Consider what I consider to be the primary scenario - x has resources to cleanup after it becomes unreachable.

x may refer to a variety of other objects; some will be the resources needing cleanup, others may be ordinary, non-resource objects.

For proper cleanup, resource objects **_already_** must remain reachable longer than x (so their cleanup can be performed even after x is unreachable). Cleaner takes care of this for you (The Cleaner instance keeps a reference to the registered "cleaning action", which refers to the resources). Or, one must roll their own if using Reference+ReferenceQueue directly.